### PR TITLE
Update nf-tests and language server fixes

### PR DESCRIPTION
**Pull request by LouisLeNezet**
- Fix image in `usage.md`.
-  Fix small warnings and errors with updated language server. 
- `def` has been added when necesary, `
- :` use instead of `,` in assertions, 
- `_` added to variables not used in closures, 
- `for` loop replaced by `.each{}`, 
- remove unused code / input. 
- Fix getFileExtension function.

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/phaseimpute/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/phaseimpute _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core pipelines lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: conf/test_full.config [hunk 1]
```diff
@@ -35,10 +35,10 @@ params {
 
     // Panelprep optional args
     remove_samples = "NA12878,NA12891,NA12892"
-    normalize      = true
-    compute_freq   = true
-    phase          = true
+    normalize      = false
```
Comment (atrigila) [inline conf/test_full.config:38]: I am not sure if this will work without normalizing. Test with full chromosome test before full test.

File: conf/test_full.config [hunk 2]
```diff
@@ -35,10 +35,10 @@ params {
...
+    compute_freq   = false
+    phase          = false
 
     // Impute tools
-    tools = "glimpse1"
+    tools = "glimpse2"
```
Comment (atrigila) [inline conf/test_full.config:43]: Why this change?

**Review by atrigila — CHANGES_REQUESTED** [review]
- Small suggestions to document the reason for the changes in the PR. 
- Add information to changelog. 
- Test (and post results) of the full chromosome test before the full test.

File: conf/test_full.config (hunk 2 above)
Comment (LouisLeNezet) [reply conf/test_full.config:43]: It is to reduce the computational time and cost, as glimpse1 has more step and takes more time to compute.

File: modules/local/vcf_chr_extract/main.nf
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply modules/local/vcf_chr_extract/main.nf:1]: Done

File: subworkflows/local/bam_chr_rename_samtools/main.nf
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply subworkflows/local/bam_chr_rename_samtools/main.nf:1]: Done

File: subworkflows/local/bam_impute_stitch/main.nf
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply subworkflows/local/bam_impute_stitch/main.nf:1]: The changes here are made as `ngen` and  `buffer` are also use as variable in the closure below.

File: subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf [hunk 3]
```diff
@@ -351,40 +351,40 @@ workflow PIPELINE_COMPLETION {
 def validateInputParameters() {
     genomeExistsError()
     // Check that only genome or fasta is provided
-    assert params.genome == null || params.fasta == null, "Either --genome or --fasta must be provided"
-    assert !(params.genome == null && params.fasta == null), "Only one of --genome or --fasta must be provided"
+    assert (params.genome == null || params.fasta == null) : "Either --genome or --fasta must be provided"
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf:354]: Yeah indeed, I've added to the changelog.

File: subworkflows/local/utils_nfcore_phaseimpute_pipeline/tests/function.nf.test.snap
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply subworkflows/local/utils_nfcore_phaseimpute_pipeline/tests/function.nf.test.snap:1]: I've added the fixing of the getFileExtension in the Changelog

File: conf/test_full.config (hunk 1 above)
Comment (LouisLeNezet) [reply conf/test_full.config:38]: Effectively to remove the samples from the panel reference we need the normalisation step.
I will add it back.

File: conf/test_full.config
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply conf/test_full.config:1]: I'm testing it on 4 chromosomes on my HPC

[the remark this answers is not in the document]
Comment (atrigila) [reply conf/test_full.config:1]: Cool, as soon as you share those results and the pipeline works well we can merge this PR :)

[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply conf/test_full.config:1]: My cluster is really slow and it takes a long time to compute (it does one job at a time 😢 ).
But everything is working properly now.

[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply conf/test_full.config:1]: I've asked on slack if we should do the fulltest on a whole genome or on a subset of chromosome to reduce the cost even more.

File: workflows/phaseimpute/tests/main.nf.test.snap [hunk 4]
```diff
@@ -232,17 +232,17 @@
...
-                "NA20359"
+                "NA12878",
+                "NA19401"
             ],
             "VcfFile [chromosomes=[chr21, chr22], sampleCount=2, variantCount=1739, phased=true]",
             [
-                "NA19401",
-                "NA20359"
+                "NA12878",
+                "NA19401"
```
Comment (atrigila) [inline workflows/phaseimpute/tests/main.nf.test.snap:241]: Why these changed?

File: conf/test_all.config [hunk 5]
```diff
@@ -33,6 +33,7 @@ params {
     phase        = true
     normalize    = true
     compute_freq = false
+    chunk_model  = "recursive"
```
Comment (atrigila) [inline conf/test_all.config:36]: What is the rationale behind this? Why adding this now and in this PR? We shouldn't be making a lot of new changes or additions before the first release PR. We are supposed to correct the issues/changes required by reviewer. The problem with adding new features now is that:
- Increases the scope of this PR, making it difficult for me to review all new changes all the time. 
- We would have to test all of these new features, add nf-tests, etc. 

If the full chromosome test passed without these changes, then it is fine, we shouldn't add any new features (at least in this PR or stage).

Comment (LouisLeNezet) [reply conf/test_all.config:36]: This is due to the fulltest not passing due to an overlapp in the chunks.
I've got an error at the ligation step that said:

`Three files overlapping at position: 10804701`

I've looked into the issue of the Glimpse2 repository and the advice was maybe to change the chunk model (we first went with `--recursive` as `--sequential` was not working on small region) and to increase the minimal window size.

File: workflows/phaseimpute/tests/main.nf.test.snap (hunk 4 above)
Comment (LouisLeNezet) [reply workflows/phaseimpute/tests/main.nf.test.snap:241]: I didn't updated it before after fixing the sample id sorting.
This is now in the right order.

File: conf/test_all.config (hunk 5 above)
Comment (LouisLeNezet) [reply conf/test_all.config:36]: I've launched back the test on my HPC but it's really slow as my HPC is full.

Comment (atrigila) [reply conf/test_all.config:36]: I would say to stick to glimpse1 then, which we know was working for the full chromosome test. Then in a **separate** PR we can review if changes to glimpse2 are necessary.

Comment (LouisLeNezet) [reply conf/test_all.config:36]: I can create a separate PR for this if you prefer.

File: conf/test_full.config
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply conf/test_full.config:1]: This should be removed now

File: workflows/phaseimpute/tests/main.nf.test.snap (hunk 4 above)
Comment (atrigila) [reply workflows/phaseimpute/tests/main.nf.test.snap:241]: But not only the order changed, the named did. Before it was: NA19401, **NA20359** and now it is **NA12878**, NA19401.

**Review by atrigila — CHANGES_REQUESTED** [review]
Thanks for the PR. There’s a lot of great work in here. I noticed, though, that the scope of this PR is still a bit broad. Since the main focus is on "reducing computational time for the AWS megatest," could we keep this PR to just those specific changes? Some of the updates seem unrelated, and reviewing all 64 files at once makes it a bit tough to keep things focused.

Would you mind splitting out the unrelated changes into separate PRs (e.g. subworkflow snapshots, major language server fixes)? I'll make it better for me to give each code the attention it deserves. Also, include the results of at least one full chromosome test (you can run it on a large gitpod instance with a small sample and it shouldn't take that long) showing the performance improvement will be super helpful to speed up the review.

File: subworkflows/local/bam_chr_rename_samtools/tests/main.nf.test [hunk 6]
```diff
@@ -42,11 +42,17 @@ nextflow_workflow {
         then {
             assertAll(
                 { assert workflow.success },
-                { assert snapshot(workflow.out).match() },
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply subworkflows/local/bam_chr_rename_samtools/tests/main.nf.test:45]: Yes due to timestamp in the vcf

File: workflows/phaseimpute/tests/main.nf.test.snap (hunk 4 above)
Comment (LouisLeNezet) [reply workflows/phaseimpute/tests/main.nf.test.snap:241]: Yeah as the order should be NA12878, NA19401, NA20359 then split in a first file of 2 and a file of 1 sample.

**Comment by LouisLeNezet** [comment]
The scope of the PR has been changed to only fix the language server and nf-test updates.
The fulltest changes has been moved to #158

**Review by atrigila — APPROVED** [review]
